internal/pipeline/parser.go

```diff
@@ -1,184 +1,86 @@
 package pipeline
 
 import (
-	"bytes"
 	"errors"
-	"fmt"
+	"io"
 	"strings"
 
 	"github.com/buildkite/agent/v3/env"
-	"github.com/buildkite/agent/v3/internal/yamltojson"
+	"github.com/buildkite/agent/v3/internal/ordered"
 	"github.com/buildkite/agent/v3/tracetools"
 	"github.com/buildkite/interpolate"
 
 	"gopkg.in/yaml.v3"
 )
 
-// Parser parses a pipeline, optionally interpolating values from
-// a given environment.
-type Parser struct {
-	Env             *env.Environment
-	Filename        string
-	Pipeline        []byte
-	NoInterpolation bool
-}
-
-// Parse runs the parser.
-func (p *Parser) Parse() (*ParserResult, error) {
-	if p.Env == nil {
-		p.Env = env.New()
-	}
-
-	errPrefix := "Failed to parse pipeline"
-	if p.Filename != "" {
-		errPrefix = fmt.Sprintf("Failed to parse %s", p.Filename)
-	}
-
-	// Run the pipeline through the YAML parser. Parse to a *yaml.Node because
-	// that will accept anything, but means we have more work to do.
-	var doc yaml.Node
-	if err := yaml.Unmarshal(p.Pipeline, &doc); err != nil {
-		return nil, fmt.Errorf("%s: %v", errPrefix, formatYAMLError(err))
-	}
-
-	// Some quick paranoia checks...
-	// yaml.v3 parses the top level as a DocumentNode, and only parses one child
-	// node (that then has the content we care about).
-	if doc.Kind != yaml.DocumentNode {
-		// TODO: Use %v once yaml.Kind has a String method
-		return nil, fmt.Errorf("%s: line %d, col %d: pipeline is not a YAML document? (kind = %x)", errPrefix, doc.Line, doc.Column, doc.Kind)
-	}
-	if len(doc.Content) != 1 {
-		return nil, fmt.Errorf("%s: line %d, col %d: pipeline document contains %d top-level nodes, want 1", errPrefix, doc.Line, doc.Column, len(doc.Content))
-	}
-
-	// It's more useful to deal with the top-level mapping node than the
-	// document node.
-	pipeline := doc.Content[0]
-
-	// We support top-level arrays of steps. If the document content is
-	// a sequence node, hoist that out into a mapping node under the key "steps"
-	// to make it look like a modern pipeline.
-	if pipeline.Kind == yaml.SequenceNode {
-		steps := pipeline
-		pp, err := yamltojson.UpsertItem(nil, "steps", steps)
-		if err != nil {
-			return nil, fmt.Errorf("%s: %v", errPrefix, err)
-		}
-		pipeline = pp
+// Parse parses a pipeline and applies interpolation. If nil envMap is provided,
+// interpolation is skipped. Otherwise, Parse mutates envMap to contain
+// variables from the pipeline's env block.
+// To interpolate only using variables defined the pipeline, pass an empty but
+// non-nil envMap.
+func Parse(src io.Reader, envMap *env.Environment) (*Pipeline, error) {
```

Comment:
i wonder if from a defensive programming perspective, the difference between an empty env and a nil env is a bit slim... perhaps we could split this into two methods:
```Go
func Parse(src io.Reader) (*Pipeline, error)
func (p *Pipeline) Interpolate(envMap *env.Environment) error
```

and then let the caller handle the decision of whether or not to interpolate.

Reply:
Done!